src/QsCompiler/Tests.Compiler/TestCases/CycleDetection.qs

```diff
@@ -112,4 +112,75 @@ namespace Microsoft.Quantum.Testing.CycleDetection {
         Bar();
         Baz();
     }
+}
+
+// =================================
+
+// Johnson's Graph Cycles
+namespace Microsoft.Quantum.Testing.CycleDetection {
```

Comment:
With a single complicated test like this is hard to analyse the coverage and corner cases of the algorithm. While it's worth keeping it, I'd add a few smaller test cases that would also illustrate to the reader some of the concerns the algorithm has to address. For example:
1. no cycles
2. two independent elementary cycles in different SCCs
3. function that calls itself
4. two cycles that have exactly one vertex in common
5. two cycles that have two vertices in common (thus producing 4 different elementary cycles)
6. "sausage link" graph {a->b, b->a; b->c, c->b; c->d, d->c} -- in this example only three elementary cycles

Reply:
There are other tests in the file that meet many of your examples. This one is called the Johnson's Graph because it is an example that came from the paper. I like the idea of the "sausage link" test and will add that. I will also add something for your fifth case.

Reply:
I'd still suggest to add a dedicated test for the case of two separate SCCs with a cycle in each (it's probably covered in the mega-test but I find it hard to analyze that one)